### PR TITLE
Optimizations for attention policy head

### PR DESCRIPTION
**Pull request by ankan-ban**
1. Do keys, queries, values matrix multiplications in parallel in the encoder layers (as batched GEMM).
2. Do keys, queries matrix multiplication for policy head in parallel.
3. Add kernel to do batched bias addition. Optimize it to reach close to full memory bandwidth utilization (AddVectors kernel was not written with performance in mind).
4. Optimize LayerNorm and Softmax kernels to reach close to full memory bandwidth utilization.

About 7% improvement in total end to end network evaluation time on RTX 3080 for the 20b Attention head network I got from Arcturai (20b_43-swa-2000000.pb.gz):
Before: Benchmark batch size 256 with inference average time 9.68394ms - throughput 26435.5 nps.
After: Benchmark batch size 256 with inference average time 9.07259ms - throughput 28216.8 nps.

### CODE REVIEW

File: src/neural/cuda/fp16_kernels.cu [hunk 1]
```diff
@@ -443,7 +443,7 @@ void OutputInputTransform(int N, int C, int se_K, T* output, const T* input,
           72 * C * sizeof(half));
       OutputInputTransformKernel_fp16_shmem_board<activation, use_bias,
                                                   use_skip>
-          <<<N, C, kMaxResBlockFusingSeFp16AmpereSmem, stream>>>(
+          <<<N, C, 72 * C * sizeof(half), stream>>>(
```
Comment (Tilps) [inline src/neural/cuda/fp16_kernels.cu:446]: Stray diff?

File: src/neural/cuda/network_cuda.cc [hunk 2]
```diff
@@ -76,7 +76,10 @@ static size_t getMaxAttentionHeadSize(const LegacyWeights& weights, int N) {
   const size_t output_size = N * (64 * 64 + 8 * 24);
 
   size = std::max(size, std::max(matmul_qk_size, output_size));
-  return size;
+
+  // We process the k, q, and v matrix multiplications in parallel using single allocation
+  // so need 3x of this.
+  return size * 3;
```
Comment (Tilps) [inline src/neural/cuda/network_cuda.cc:82]: This seems pessimistic - shouldn't you multiply encoder_d_model by 3 instead and policy_d_model by 2?

File: src/neural/cuda/network_cuda.cc [hunk 3]
```diff
@@ -280,10 +283,9 @@ class CudaNetwork : public Network {
     scratch_size_ = std::max(scratch_size_, 2 * transformed_tensor_size);
 
     // Attention policy head may need more memory
-    // (We also split the allocations into two parts, so need 2x)
     const size_t attentionSize =
         getMaxAttentionHeadSize(weights, max_batch_size_);
-    scratch_size_ = std::max(scratch_size_, 2 * attentionSize);
+    scratch_size_ = std::max(scratch_size_, attentionSize);
```
Comment (Tilps) [inline src/neural/cuda/network_cuda.cc:288]: hrmm not sure about this. Allocations are still split?

File: src/neural/cuda/fp16_kernels.cu (hunk 1 above)
Comment (ankan-ban) [reply src/neural/cuda/fp16_kernels.cu:446]: No, it's actually a fix that I forgot to include with my last change (otherwise nets bigger than 384 filters and smaller than 512 filters would fail here).

File: src/neural/cuda/network_cuda.cc (hunk 2 above)
Comment (ankan-ban) [reply src/neural/cuda/network_cuda.cc:82]: yeah, will make it more accurate.

File: src/neural/cuda/network_cuda.cc (hunk 3 above)
Comment (ankan-ban) [reply src/neural/cuda/network_cuda.cc:288]: I added "size * 3" in getMaxAttentionHeadSize() which should take care of this part (It's not very obvious, but the code doesn't actually use second part of the allocation when using them for batched gemms)

Comment (Tilps) [reply src/neural/cuda/network_cuda.cc:288]: If getMaxAttentionHeadSize uses multipliers for each phase as appropriate - depending on whether they use the bottom half only or top as well as well as the actual size after batching - then yes it makes sense to drop the 2 here.

Comment (ankan-ban) [reply src/neural/cuda/network_cuda.cc:288]: I have simplified intermediate memory allocations a bit (also reduced from 5 scratch allocations to 4). Hopefully it's accurate now.

File: src/neural/cuda/network_cuda.cc [hunk 4]
```diff
@@ -68,14 +68,17 @@ static size_t getMaxAttentionHeadSize(const LegacyWeights& weights, int N) {
...
   size_t size = N * 64 *
-                std::max(std::max(embedding_op_size, encoder_dff),
-                         std::max(policy_d_model, encoder_d_model));
+      std::max(std::max(embedding_op_size, encoder_dff), policy_d_model);
 
   // size of matmul_qk matrix = encoder_heads_ * Batch * 64 * 64
   const size_t matmul_qk_size = encoder_heads * N * 64 * 64;
   const size_t output_size = N * (64 * 64 + 8 * 24);
-
   size = std::max(size, std::max(matmul_qk_size, output_size));
```
Comment (Tilps) [inline src/neural/cuda/network_cuda.cc:76]: output_size goes in a single allocation right? - so since we're going to multiply by 2 down below, maybe it could be halved? (Or maybe just don't bother :P)

File: src/neural/cuda/layers.cc [hunk 5]
```diff
@@ -1677,7 +1691,7 @@ void AttentionPolicyHead<DataType>::Eval(
   DataType* promotion_logits = output + 64 * 64;
 
   ComputePromotionLogits<DataType>(N, policy_d_model_, promotion_logits,
-                                   scratch2, ip4_pol_w_, output, stream);
+                                   scratch0, ip4_pol_w_, output, stream);
```
Comment (Tilps) [inline src/neural/cuda/layers.cc:1694]: There was a bug report that promotions are broken with this PR.  Looking at the code I think that scratch0 here should actually be wk.

Comment (Tilps) [reply src/neural/cuda/layers.cc:1694]: Kovax tried out this fix and it seemed to work.

Comment (ankan-ban) [reply src/neural/cuda/layers.cc]: sorry about that! Yeah, I can confirm that was unintentional.

**Review by Tilps — APPROVED** [review]
